### PR TITLE
[`BC 4.37 -> 4.38`] for Llama family, memory and speed 

### LINKED ISSUE #29412: Codellama-7b is 3x slower in 4.38.2 compared to 4.34.0 with CPU backend
### System Info

transformers 4.34.0 is ~370ms/token while 4.38.2 it is ~990 ms/token. Model size is also slightly larger in 4.38.2.

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction
```python
inputs = tokenizer(" def fibonacci_recursive(n):", return_tensors="pt")
attention_mask = torch.ones(inputs.input_ids.shape)
start = time.perf_counter()
generate_ids = model.generate(input_ids=inputs.input_ids, attention_mask=attention_mask, max_new_tokens=100)
end = time.perf_counter()
print(f"{end-start")
``` 

### Expected behavior

I expect the model/weights to be the same and produce same latency. 
This discrepancy is not observed in Llama-2-7B though.

### LINKED ISSUE #29484: Upgrade from 4.37.2 to 4.38.2 causes CUDA out of memory error with identical configuration.
### System Info

- `transformers` version: 4.38.2
- Platform: Linux-5.15.0-88-generic-x86_64-with-glibc2.31
- Python version: 3.9.18
- Huggingface_hub version: 0.21.3
- Safetensors version: 0.4.2
- Accelerate version: 0.27.2
- Accelerate config:    not found
- PyTorch version (GPU?): 2.2.1
- Tensorflow version (GPU?): not installed
- Flax version (CPU?/GPU?/TPU?): not installed
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: Yes

### Who can help?

@muellerz 
@pacman100
@gante 
@Narsil 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I was using the official training script provided by DeepSeek-Coder (https://github.com/deepseek-ai/DeepSeek-Coder/blob/main/finetune/finetune_deepseekcoder.py) to train a model on the dataset (https://huggingface.co/datasets/nickrosh/Evol-Instruct-Code-80k-v1) when I noticed that using transformers version 4.38.2 results in an error: torch.cuda.OutOfMemoryError: CUDA out of memory. However, simply downgrading transformers to version 4.37.2 resolves this issue.

The command line I used was as follows:
```
#!/bin/bash
export DS_SKIP_CUDA_CHECK=1

DATA_PATH="/data2/user/llm/Evol-Instruct-Code-80k-v1/EvolInstruct-Code-80k.json"
OUTPUT_PATH="/data2/user/llm/sft/0306"
MODEL_PATH="/data2/user/llm/deepseek-coder-1.3b-base"

export WANDB_PROJECT="llm"
export WANDB_LOG_MODEL="false"
export WANDB_WATCH="false"

deepspeed --include localhost:0,1,2,3,4,5,6,7 DeepSeek-Coder/finetune/finetune_deepseekcoder.py \
    --model_name_or_path $MODEL_PATH \
    --data_path $DATA_PATH \
    --output_dir $OUTPUT_PATH \
    --num_train_epochs 3 \
    --model_max_length 1024 \
    --per_device_train_batch_size 64 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --evaluation_strategy "no" \
    --save_strategy "epoch" \
    --save_steps 100 \
    --save_total_limit 100 \
    --learning_rate 2e-5 \
    --warmup_steps 10 \
    --lr_scheduler_type "cosine" \
    --gradient_checkpointing True \
    --report_to "wandb" \
    --logging_steps 1 \
    --run_name "test"
    --deepspeed DeepSeek-Coder/finetune/configs/ds_config_zero3.json \
    --bf16 True \
    --seed 42
```

The details error information can be found as:
```
Traceback (most recent call last):
  File "/data2/user/llm/DeepSeek-Coder/finetune/finetune_deepseekcoder.py", line 200, in <module>
    train()
  File "/data2/user/llm/DeepSeek-Coder/finetune/finetune_deepseekcoder.py", line 194, in train
    trainer.train()
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/trainer.py", line 1624, in train
    return inner_training_loop(
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/trainer.py", line 1961, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/trainer.py", line 2902, in training_step
    loss = self.compute_loss(model, inputs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/trainer.py", line 2925, in compute_loss
    outputs = model(**inputs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1523, in forward
    else self._run_ddp_forward(*inputs, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1359, in _run_ddp_forward
    return self.module(*inputs, **kwargs)  # type: ignore[index]
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 1176, in forward
    outputs = self.model(
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 993, in forward
    causal_mask = self._update_causal_mask(attention_mask, inputs_embeds)
  File "/home/user/anaconda3/envs/llm/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 1093, in _update_causal_mask
    causal_mask = causal_mask.mul(~torch.all(causal_mask == min_dtype, dim=-1, keepdim=True)).to(dtype)
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 32.00 GiB. GPU 6 has a total capacity of 79.15 GiB of which 8.24 GiB is free. Including non-PyTorch memory, this process has 70.91 GiB memory in use. Of the allocated memory 38.74 GiB is allocated by PyTorch, and 31.00 GiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
```

I noticed that when this error occurs, the monitor indeed shows that there is not enough space on the GPU. However, at this time, only the training program is running on the server. This phenomenon seems to imply that while the program is occupying a large amount of VRAM, it is still requesting more VRAM space, a problem that was not encountered in the previous version of the transformers.

Given my repeated experiments, I believe we can essentially rule out the script or parameter settings as the source of the issue. Since I've utilized DeepSpeed to accelerate training, I personally think that this problem might occur during the parallel training process. Particularly, the code changes introduced by this major version update seem likely to be a contributing factor to the issue I'm encountering.

### Expected behavior

First of all, I would like to express my gratitude to the developers and maintainers for their continuous effort in maintaining and improving this repository. Your hard work is truly appreciated by the community.
Regarding the expected behavior, it would be for the model to train successfully without encountering a torch.cuda.OutOfMemoryError, leveraging GPU memory as efficiently as it did in the previous transformers library version (version 4.37.2). The ideal scenario involves the program dynamically managing memory allocation to prevent the depletion of GPU resources, thus ensuring a stable and efficient training process. This level of performance was observed in the past and would be anticipated to continue without necessitating manual adjustments or a downgrade to an earlier version.
Thanks for you reviewing again.

### LINKED ISSUE #29644: Llama runs very slow in 4.38.2
### System Info

- `transformers` version: 4.38.2
- Platform: Linux-4.18.0-513.5.1.el8_9.x86_64-x86_64-with-glibc2.28
- Python version: 3.10.13
- Huggingface_hub version: 0.20.2
- Safetensors version: 0.4.1
- Accelerate version: 0.26.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: NVIDIA A100 GPUs
- Using distributed or parallel set-up in script?: "auto"

### Who can help?

@ArthurZucker @gante

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

pip install transformers==4.38.2

### Expected behavior

When I use transformers==4.36.2, Llama 2 functions well. However, after updating to 4.38.2, Llama runs very much slower (20 times). No error, no warning, just very slow running.

Please help to check the issue on a timely basis! Thanks!

### LINKED ISSUE #29651: output decoding perf. of transfomers==4.38.2
There's some perf gap between 4.38.2 vs 4.35.0 during output decoding. The LlamaAttention was used. (https://github.com/huggingface/transformers/blob/f738ab3b5d30e30c43a4c3d00ca8939f8a4d4427/src/transformers/models/llama/modeling_llama.py#L688)

Is it something related to (larger) static kv cache movement feature in 4.38.2 during output decoding ? 

Regards, 
==
```
**Nvidia A100**						
	[INFO] torch:   2.3.0.dev20240313+cu121					
	[INFO] cuda/rocm:    12010					
	[INFO] device:  NVIDIA A100 80GB PCIe					
	**[INFO] transformers:  4.35.0**					
	[INFO] model:  TheBloke/Llama-2-7B-Chat-fp16  -head_num:  32  --num_key_value_heads:  32  -head_dim:  128  -layers_num: 32					

	       batch_size:  1  seq_len:  1  output_tokens:  32			       Avg per token latency:        33.41 ms		
	       batch_size:  8  seq_len:  1  output_tokens:  32			       Avg per token latency:        34.18 ms		
	       batch_size:  16  seq_len:  1  output_tokens:  32			       Avg per token latency:        34.28 ms		
	       batch_size:  32  seq_len:  1  output_tokens:  32			       Avg per token latency:        34.43 ms	
					
	[INFO] torch:   2.3.0.dev20240313+cu121					
	[INFO] cuda/rocm:    12010					
	[INFO] device:  NVIDIA A100 80GB PCIe					
	**[INFO] transformers:  4.38.2**					
	[INFO] model:  TheBloke/Llama-2-7B-Chat-fp16  -head_num:  32  --num_key_value_heads:  32  -head_dim:  128  -layers_num: 32					

	       batch_size:  1  seq_len:  1  output_tokens:  32			       Avg per token latency:        38.12 ms		
	       batch_size:  8  seq_len:  1  output_tokens:  32			       Avg per token latency:        39.97 ms		
	       batch_size:  16  seq_len:  1  output_tokens:  32			       Avg per token latency:        41.29 ms		
	       batch_size:  32  seq_len:  1  output_tokens:  32			       Avg per token latency:        44.05 ms	

**AMD MI210**						
	[INFO] torch:   2.3.0.dev20240313+rocm6.0					
	[INFO] cuda/rocm:    60000					
	[INFO] device:  AMD Instinct MI210					
	**[INFO] transformers:  4.35.0**					
	[INFO] model:  TheBloke/Llama-2-7B-Chat-fp16  -head_num:  32  --num_key_value_heads:  32  -head_dim:  128  -layers_num: 32					

	       batch_size:  1  seq_len:  1  output_tokens:  32			       Avg per token latency:        33.43 ms		
	       batch_size:  8  seq_len:  1  output_tokens:  32			       Avg per token latency:        35.07 ms		
	       batch_size:  16  seq_len:  1  output_tokens:  32			       Avg per token latency:        35.59 ms		
	       batch_size:  32  seq_len:  1  output_tokens:  32			       Avg per token latency:        39.15 ms		
						
	[INFO] torch:   2.3.0.dev20240313+rocm6.0					
	[INFO] cuda/rocm:    60000					
	[INFO] device:  AMD Instinct MI210					
	**[INFO] transformers:  4.38.2**					
	[INFO] model:  TheBloke/Llama-2-7B-Chat-fp16  -head_num:  32  --num_key_value_heads:  32  -head_dim:  128  -layers_num: 32					

	       batch_size:  1  seq_len:  1  output_tokens:  32			       Avg per token latency:        35.66 ms		
	       batch_size:  8  seq_len:  1  output_tokens:  32			       Avg per token latency:        37.81 ms		
	       batch_size:  16  seq_len:  1  output_tokens:  32			       Avg per token latency:        40.44 ms		
	       batch_size:  32  seq_len:  1  output_tokens:  32			       Avg per token latency:        48.14 ms
```

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Fixes the BC issues between the two versions in term of memory consumption.
This fix is made a lot easier by all the tests, so thanks a lot @gante! 

fixes #29412, fixes #29484 , fixes #29644, fixes #29651

### CODE REVIEW

**Review by gante — COMMENTED** [review]
In general looks good to me, although I'm not 100% sure on the `causal_mask *= torch.arange(target_length, device=device) > cache_position[0]` line + assisted generation -- going to have a look

**Comment by ArthurZucker** [comment]
torch script can be fixed by @fxmarty in a follow up PR + patch IMO!

**Comment by fxmarty** [comment]
@ArthurZucker that would be great if torchscript/fx tests pass &

`optimum-cli export onnx --model fxmarty/tiny-llama-fast-tokenizer llama_onnx`

does not break

File: src/transformers/models/cohere/modeling_cohere.py [hunk 1]
```diff
@@ -559,8 +557,9 @@ def forward(
         value_states = repeat_kv(value_states, self.num_key_value_groups)
 
         causal_mask = attention_mask
-        if attention_mask is not None and cache_position is not None:
-            causal_mask = causal_mask[:, :, cache_position, : key_states.shape[-2]]
+        # if attention_mask is not None and cache_position is not None:
+        if attention_mask is not None:
```
Comment (younesbelkada) [inline src/transformers/models/cohere/modeling_cohere.py:561]:
```suggestion
        if causal_mask is not None:
```

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -658,8 +656,9 @@ def forward(
         value_states = repeat_kv(value_states, self.num_key_value_groups)
 
         causal_mask = attention_mask
-        if attention_mask is not None and cache_position is not None:
-            causal_mask = causal_mask[:, :, cache_position, : key_states.shape[-2]]
+        # if attention_mask is not None and cache_position is not None:
+        if attention_mask is not None:
```
Comment (younesbelkada) [inline src/transformers/models/llama/modeling_llama.py:660]:
```suggestion
        if causal_mask is not None:
```

**Review by younesbelkada — APPROVED** [review]
Thanks for the offline explanation ! this shouldn't affect FA2 as we always return `attention_mask` without processing it for FA2 modules in `_update_causal_mask`, as you explained offline `causal_mask *= torch.arange(target_length, device=device) > cache_position[0]` is used to mask out the cached hidden states

**Comment by gante** [comment]
@ArthurZucker the line `causal_mask *= torch.arange(target_length, device=device) > cache_position[0]` should become `causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)`

In the previous version, in a situation with 17 cached tokens and 4 new assistant tokens, the causal mask would become
```
(Pdb) causal_mask
tensor([[[[     0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0., -65504., -65504., -65504., -65504.],
          [     0.,      0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0., -65504., -65504., -65504., -65504.],
          [     0.,      0.,      0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0., -65504., -65504., -65504., -65504.],
          [     0.,      0.,      0.,      0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0., -65504., -65504., -65504., -65504.],
          [     0.,      0.,      0.,      0.,      0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0., -65504., -65504., -65504., -65504.]]]],
       device='cuda:0', dtype=torch.float16)
```

i.e. not upper triangular. After the suggested change, it becomes

```
(Pdb) causal_mask
tensor([[[[     0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0., -65504., -65504., -65504., -65504.],
          [     0.,      0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0., -65504., -65504., -65504.],
          [     0.,      0.,      0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0., -65504., -65504.],
          [     0.,      0.,      0.,      0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0., -65504.],
          [     0.,      0.,      0.,      0.,      0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.,
               -0.,     -0.,     -0.,     -0.,     -0.,     -0.,     -0.]]]],
       device='cuda:0', dtype=torch.float16)
```

**Comment by ArthurZucker** [comment]
Any idea why our tests don't complain?

**Comment by gante** [comment]
> Any idea why our tests don't complain?

I think we don't have hard correctness checks for assisted generation 🙈 only API checks

**Comment by poedator** [comment]
congratulations with this PR fixing many important things!
Unfortunately, it broke ~~my heart~~ custom 4D masks support. 
try `RUN_SLOW=1 python -m pytest -v ./tests/test_modeling_utils.py::Mask4DTestHard::test_partial_stacked_causal_mask`

@gante and myself introduced this test recently in #29731 - please make it a part of the test suite for all things related to attention_masks and `StaticCache`

here is what happens (numbers based on `test_partial_stacked_causal_mask` after [this line 2170](https://github.com/gante/transformers/blob/a97a7e907978d1e08c45bdffbdab1b99358dc712/tests/test_modeling_utils.py#L2170):
in `modeling_llama.py::_update_causal_mask()`
the `causal_mask` has shape `(1, 1, sequence_length, target_length)` (1,1,9,12) (3 tokens in cache)
if the custom 4D mask enters with same shape, it triggers offset = 3 and then causes error when copied over causal mask.
if the custom 4D mask enters with shape (1, 1, 12, 12), offset stays at zero but then again this causes error when copied over causal mask, for mask_slice and causal_mask shapes don't match.
There are also changes [here](https://github.com/huggingface/transformers/blob/1abd0986ddde1b7235e691703c32cd8bda3ae069/src/transformers/models/llama/modeling_llama.py#L661) which may affect this test.

I hesitate to offer a PR because it may break other things that you try to do with this part of code. But, please, make the test work. BTW, it may be OK to change the test, for instance passing the whole bigger mask including cached items by editing [this line](https://github.com/gante/transformers/blob/a97a7e907978d1e08c45bdffbdab1b99358dc712/tests/test_modeling_utils.py#L2168) to `mask_1b = mask_1`

Special note on StaticCache: I like this feature and I want to use custom 4D masks with it. So far this is not tested. I'd be glad to contribute such test once this issue is fixed. It will look like `test_partial_stacked_causal_mask`, only with StaticCache.

cc @ArthurZucker @gante

**Comment by ArthurZucker** [comment]
I can try to fix it, and yes I thought it would be tested automatically. It should be part of the `tokenization_common` or at least LlamaIntegrationTests or something. cc @gante if you can take a look I'll gladly review a PR

**Comment by gante** [comment]
^ this PR should fix it 🤗

**Pull request merged**